include/mxnet/c_api.h

```diff
@@ -737,6 +741,57 @@ MXNET_DLL int MXNDArrayGetShape(NDArrayHandle handle,
  */
 MXNET_DLL int MXNDArrayGetData(NDArrayHandle handle,
                                void **out_pdata);
+/*!
+* \brief Create a reference view of NDArray that
+*  represents as DLManagedTensor until
+*  all the pending writes with respect NDArray are finished.
+* \param handle the handle to the ndarray
+* \param out_dlpack pointer holder to get pointer of DLManagedTensor
+* \return 0 when success, -1 when failure happens
+*/
+MXNET_DLL int MXNDArrayToDLPackForRead(NDArrayHandle handle,
+                                       DLManagedTensorHandle *out_dlpack);
+
+/*!
+* \brief Create a reference view of NDArray that
+*  represents as DLManagedTensor until
+*  all the pending reads/writes with respect NDArray are finished.
+* \param handle the handle to the ndarray
+* \param out_dlpack pointer holder to get pointer of DLManagedTensor
+* \return 0 when success, -1 when failure happens
+*/
+MXNET_DLL int MXNDArrayToDLPackForWrite(NDArrayHandle handle,
+                                        DLManagedTensorHandle *out_dlpack);
+
+/*!
+* \brief Create a NDArray backed by a dlpack tensor.
+*
+* This allows us to create a NDArray using the memory
+* allocated by an external deep learning framework
+* that is DLPack compatible.
+*
+* The memory is retained until the NDArray went out of scope.
+*
+* \param dlpack the pointer of the input DLManagedTensor
+* \param out_handle pointer holder to get pointer of NDArray
+* \return 0 when success, -1 when failure happens
+*/
+MXNET_DLL int MXNDArrayFromDLPack(DLManagedTensorHandle dlpack,
+                                  NDArrayHandle *out_handle);
+/*!
+ * \brief Delete a dlpack tensor
+ * \param dlpack the pointer of the input DLManagedTensor
+ * \return 0 when success, -1 when failure happens
+ */
+MXNET_DLL int MXNDArrayCallDLPackDeleter(DLManagedTensorHandle dlpack);
+
+/*!
+ * \brief Delete a dlpack tensor
+ * \param dlpack_capsule the pointer of a PyCapsule storing DLManagedTensor
+ * \return 0 when success, -1 when failure happens
+ */
+MXNET_DLL void MXNDArrayCallDLPackCapsuleDeleter(PyObjectHandle dlpack_capsule);
```

Comment:
I am less certain why we need the deleter function here, can they be directly handled in the python/cython side?

Reply:
I tried to implement a deleter function in python, however the deleter function may be released by Python GC before calling the deleter function.
The Python Frontend of MXNet both uses [Python(ctypes)](https://github.com/apache/incubator-mxnet/tree/master/python/mxnet/_ctypes) and [Cython](https://github.com/apache/incubator-mxnet/tree/master/python/mxnet/cython). It may be impossible to implement the deleter function in ctypes. 
So the deleter function should be implemented in C++.

Reply:
Take a look at destructor at https://github.com/dmlc/tvm/pull/1573

Reply:
There is some subtlty here but they can never-the-less be implemented

Reply:
I ever tried to write a python function as the destructor, but it can't pass CI.
Please see [the MXNet CI result](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-12047/8/pipeline)
All windows test_dlpack failed because the destructor written in Python is released before calling it.

PyTorch implemented the destructor using Python API in C++, and CuPy implemented it by cython, namely the code will be built by C++.
However, MXNet uses ctypes and cython. I couldn't find a better way to implement the destructor except writing it in MXNet C++ API.

Reply:
There is a special trick that need to be used to implement destructor in python. https://github.com/dmlc/tvm/pull/1573 implements the destructor in python and it works

Reply:
Specificialy, the signature of the destructor need to use c_void_p instead of py_object to avoid repeatitive destruction

Reply:
Yes. I knew the trick and tried it in my previous PR. But it failed in Windows Test.
[Related CI](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-12047/8/pipeline)

It seems that the CI of TVM doesn't have Windows Test so the CI is passed.

Reply:
This is strange as destructor itself sits in the global scope and should be destructed after the dltensors(which have a local scope)

Reply:
Yes.
In [the test code](https://gist.github.com/wkcn/501cd118b8ceb1608bd6c8fa968e7912), it works in Linux but failed in Windows.

Reply:
I see two problems in your particular gist you paste. 

- The destructor need to be declared in the global scope(instead of constructing when passing to the argument)
```python

cfunc = ctypes.CFUNCTYPE(None, ctypes.c_void_p)

def dfunc(dltensor):
    pycaps = ctypes.cast(dltensor, ctypes.py_object)
    pass

c_destructor = cfunc(dfunc)
c_str_dltensor = cstr("dltensor")

def test():
    a = ctypes.pythonapi.PyCapsule_New(1, c_str_dltensor, c_destructor)
test()

```

Reply:
Thank you!
I will test it on Windows.
If it works, I will update the PR.